scipy/signal/_short_time_fft.py

```diff
@@ -0,0 +1,1659 @@
+            window is generated (default). Though symmetric windows seem for
+            most applications to be more sensible, the default of a periodic
+            windows was chosen to correspond to the default of `get_window`.
+        fft_typ : 'twosided', 'centered', 'onesided', 'onesided2X'
+            Type of FFT to be used (default 'onesided').
+            See property `fft_typ` for details.
+        mfft: int | None
+            Length of the FFT used, if a zero padded FFT is desired.
+            If ``None`` (default), the length of the window `win` is used.
+        scale_to : 'magnitude', 'psd' | None
+            If not ``None`` (default) the window function is scaled, so each
+            STFT column represents  either a 'magnitude' or a power spectral
+            density ('psd') spectrum. This parameter sets the property
+            `scaling` to the same value. See method `scale_to` for details.
+        phase_shift : int | None
+            If set, add a linear phase `phase_shift` / `mfft` * `f` to each
+            frequency `f`. The default value 0 ensures that there is no phase
+            shift on the zeroth slice (in which t=0 is centered). See property
+            `phase_shift` for more details.
+
+        Examples
+        --------
+        The following instances ``SFT0`` and ``SFT1`` are equivalent:
+
+        >>> from scipy.signal import ShortTimeFFT, get_window
+        >>> nperseg = 9  # window length
+        >>> w = get_window(('gaussian', 2.), nperseg)
+        >>> fs = 128  # sampling frequency
+        >>> hop = 3  # increment of STFT time slice
+        >>> SFT0 = ShortTimeFFT(w, hop, fs=fs)
+        >>> SFT1 = ShortTimeFFT.from_window(('gaussian', 2.), fs, nperseg,
+        ...                                 noverlap=nperseg-hop)
+
+        See Also
+        --------
+        scipy.signal.get_window: Return a window of a given length and type.
+        from_dual: Create instance using dual window.
+        ShortTimeFFT: Create instance using standard initalizer.
+        """
+        win = get_window(win_param, nperseg, fftbins=not symmetric_win)
+        return cls(win, hop=nperseg-noverlap, fs=fs, fft_typ=fft_typ,
+                   mfft=mfft, scale_to=scale_to, phase_shift=phase_shift)
+
+    @property
+    def win(self) -> NDArray:
+        """Window function as real- or complex-valued 1d array.
+
+        This attribute is read only, since `dual_win` depends on it.
+
+        See Also
+        --------
+        dual_win: Canonical dual window.
+        m_num: Number of samples in window `win`.
+        m_num_mid: Center index of window `win`.
+        mfft: Length of input for the FFT used - may be larger than `m_num`.
+        hop: ime increment in signal samples for sliding window.
+        win: Window function as real- or complex-valued 1d array.
+        ShortTimeFFT: Class this property belongs to.
+        """
+        return self._win
+
+    @property
+    def hop(self) -> int:
```

Comment:
They make the user experience smoother:

- The attributes `win`, `dual_win` and `hop` are immutable, since they cannot be modified independently of each other. 
- IMO, there are applications, where it makes sense to mutate the other attributes (e.g., a GUI for an interactive spectrogram).
- The setters do parameter checking and provide friendly error messages. I feel that performing the parameter checking in separate setter methods instead of doing it all in the initializer makes the code less cluttered. 
- The mostly primitive getter methods are the obvious downside of this approach, which are not really user-visible. Hence, they are primarily a mild annoyance for the reviewers :grin: